src/weights.jl

```diff
@@ -279,6 +347,17 @@ Base.sum(v::BitArray, w::AbstractWeights) = wsum(v, values(w))
 Base.sum(v::SparseArrays.SparseMatrixCSC, w::AbstractWeights) = wsum(v, values(w))
 Base.sum(v::AbstractArray, w::AbstractWeights) = dot(v, values(w))
 
+for v in (AbstractArray{<:Number}, BitArray, SparseArrays.SparseMatrixCSC, AbstractArray)
+    @eval begin
+        function Base.sum(v::$v, w::UnitWeights)
+            if length(v) != length(w)
+                throw(DimensionMismatch("Inconsistent array dimension."))
```

Comment:
Why does `UnitWeight` have length?
And if it has to have length then why do we need to actually check it?
Is this a useful feature vs it just always being considered the right size?
Like `LinearAlgebra.I`

Reply:
That's an interesting point. Here's an example. In computing the variance, we obtain the normalization factor through `varcorrection`. It depends on the number of observations, so we need the length of the weight vector. We should also ensure that the user has not passed the wrong length lest we use the wrong normalization factor. There are ways around it (for instance, a method for `variance` that bypasses the weights), but we want to help developers avoid defining separate methods. More generally, I think that boundary checks at different points help users catch mistakes before they snowball.

Reply:
Yes, it doesn't hurt to check lengths, and you can avoid it using `@inbounds`. I'd say that in general it's quite easy to get the appropriate length since it's the length of another input argument.

Reply:
Calculating the length isn't free for general iterators.

Reply:
Yeah, but in most cases we take arrays, right? Also you can always create an object of length `typemax` if you don't care about the checks.

Reply:
Mostly isn't always. Isn't `dropmissing` after all is one of the cases of iterators that are not free to take then length of?

These checks just don't feel great to me.
What error are we catching?
And how does that error relate to weighting?

If one is checking one got the amount of data that was expected, then isn't that better done with an explicit check on the data gathering code? So a good error can be thrown?
I feel like this just adds work for the user of weights, and the best case is they get a confusing error message.

I'm all for defensive programming, but that means checking the function you are defining can work. Not just checking arbitrary things in the calling code.


> Also you can always create an object of length typemax if you don't care about the checks.

That doesn't work, does it? This check is checking the lengths are equal, not checking a `>`.

Reply:
We don't support weights with `dropmissing` currently, the main reason being that in general we would need to skip weights corresponding to missing entries (instead of naively taking weights in their order of appearance).

> That doesn't work, does it? This check is checking the lengths are equal, not checking a `>`.

Right, it doesn't work here (I was thinking about custom functions which want to take a `UnitWeights` as default weights for simplicity of implementation). I guess we could allow something like `UnitWeights(nothing)` (or maybe `UnitWeights`) to create an object with a undefined length. But I still think it's nice to take a length by default, for checking, but also because that's required to be an actual `AbstractVector` (otherwise we need to error when calling `size`).